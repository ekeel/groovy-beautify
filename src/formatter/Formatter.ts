import { CodeBlock } from "../parser/types";
import FormatRule from "./FormatRule";

export type FormatterOptions = {
  width: number;
};

export class Formatter {
  rules: FormatRule[];
  options: FormatterOptions;

  constructor(rules?: Array<typeof FormatRule>, options?: FormatterOptions) {
    this.rules = rules?.map((r) => new r(this)) || [];
    this.options = options || { width: 80 };
  }

  format(obj: CodeBlock, indent = 0): string {
    let text = "";

    if (obj) {
      const formatRule = this.rules.find((r) => r.matches(obj));
      if (obj.start !== undefined) {
        if (formatRule) {
          text += formatRule.formatStart(obj, indent);
        }
      }

      if (formatRule) {
        text += formatRule.formatChildren(obj, indent);
      }

      if (obj.end !== undefined) {
        if (formatRule) {
          text += formatRule.formatEnd(obj, indent);
        }
      }
    }
    return text;
  }
}
