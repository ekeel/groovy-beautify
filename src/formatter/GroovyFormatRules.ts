import { CodeBlock } from "../parser/Parser";
import FormatRule from "./FormatRule";

class BaseBlockRule extends FormatRule {
  beforeChild(childText: string, indent: number): string {
    let text = childText;
    const trimmedLeft = trimSpacesAndTabsLeft(text);
    if (trimmedLeft.startsWith("\n")) {
      text = "\n" + padLeft(trimmedLeft.trimStart(), indent);
    }
    const trimmedRight = trimSpacesAndTabsRight(text);
    if (trimmedRight.endsWith("\n")) {
      text = padRight(trimmedRight, indent);
    }
    return text;
  }
}

class BlockFormatRule extends BaseBlockRule {
  matches(cb: CodeBlock) {
    return cb?.type === "block";
  }

  beforeSelf(prevText: string) {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return trimmedText;
    } else {
      return prevText.trimEnd() + " ";
    }
  }

  formatEnd(cb: CodeBlock, indent: number) {
    return cb.end ? padLeft(cb.end, indent) : "";
  }

  formatChildren(cb: CodeBlock, indent: number) {
    let blockText = super.formatChildren(cb, indent + 1);
    blockText = blockText.trim();
    return "\n" + padLeft(blockText, indent + 1) + "\n";
  }
}

class InlineBlockFormatRule extends BaseBlockRule {
  matches(cb: CodeBlock) {
    return cb?.type === "round" || cb?.type === "square";
  }

  isMultiline(cb: CodeBlock) {
    return cb.children?.some((child) => child.start?.includes("\n"));
  }

  beforeSelf(prevText: string) {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return trimmedText;
    } else {
      return prevText;
    }
  }

  formatEnd(cb: CodeBlock, indent: number) {
    if (this.isMultiline(cb)) {
      return cb.end ? padLeft(cb.end, indent) : "";
    } else {
      return cb.end ?? "";
    }
  }

  formatChildren(cb: CodeBlock, indent: number) {
    if (this.isMultiline(cb)) {
      let blockText = super.formatChildren(cb, indent + 1);
      blockText = blockText.trim();
      return "\n" + padLeft(blockText, indent + 1) + "\n";
    } else {
      let blockText = super.formatChildren(cb, indent);
      return blockText.trim();
    }
  }
}

class DotSyntaxFormatRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "dot";
  }
  beforeSelf(prevText: string, indent: number): string {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return padRight(trimmedText, indent + 1);
    } else {
      return prevText;
    }
  }
}

class OperatorsRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "operators";
  }

  beforeSelf(prevText: string): string {
    return trimSpacesAndTabsRight(prevText) + " ";
  }

  afterSelf(nextText: string): string {
    return " " + nextText.trimStart();
  }

  allowBreak(): boolean {
    return true;
  }
}

class DelimitersRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "delimiters";
  }

  beforeSelf(prevText: string): string {
    return trimSpacesAndTabsRight(prevText);
  }

  afterSelf(nextText: string): string {
    return " " + trimSpacesAndTabsLeft(nextText);
  }
}

class KeywordRule extends FormatRule {
  matches(cb: CodeBlock) {
    return cb?.type === "keywords";
  }

  beforeSelf(prevText: string): string {
    const trimmedText = trimSpacesAndTabsRight(prevText);
    if (trimmedText.endsWith("\n")) {
      return trimmedText;
    } else {
      return trimmedText + " ";
    }
  }

  afterSelf(nextText: string): string {
    return " " + nextText.trimStart();
  }

  allowBreak(): boolean {
    return true;
  }
}

function padLeft(text: string, indent: number) {
  return "".padStart(indent * 4) + text;
}
function padRight(text: string, indent: number) {
  return text + "".padStart(indent * 4);
}

function trimSpacesAndTabsLeft(text: string) {
  return text.replace(/^( |\t)+/, "");
}
function trimSpacesAndTabsRight(text: string) {
  return text.replace(/( |\t)+$/, "");
}

export default [
  BlockFormatRule,
  InlineBlockFormatRule,
  DotSyntaxFormatRule,
  KeywordRule,
  OperatorsRule,
  DelimitersRule,
  FormatRule,
];
