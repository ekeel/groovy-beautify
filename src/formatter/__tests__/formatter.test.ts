import { Formatter } from "../Formatter";
import GroovyFormatRules from "../GroovyFormatRules";

test("Basic formatting", () => {
  const object = {
    type: "block",
    start: "{",
    end: "}",
    children: [
      { start: "if", type: "keywords" },
      { start: "(", end: ")", type: "round", children: [{ start: "text", type: "true" }] },
      { start: "{", end: "}", type: "block", children: [{ start: "test", type: "true" }] },
    ],
  };
  const formatter = new Formatter(GroovyFormatRules);
  const formattingResult = formatter.format(object);
  expect(formattingResult).toMatchSnapshot();
});

test("New line and tabs format", () => {
  const object = {
    type: "block",
    start: "{",
    end: "}",
    children: [
      { start: "    ", type: "text" },
      { start: "def", type: "keywords" },
      { start: "\n\t", type: "text" },
      { start: "variable", type: "identifiers" },
      { start: " ", type: "text" },
      { start: "=", type: "operators" },
      { start: "true", type: "text" },
    ],
  };
  const formatter = new Formatter(GroovyFormatRules);
  const formattingResult = formatter.format(object);
  expect(formattingResult).toMatchSnapshot();
});

export {};
