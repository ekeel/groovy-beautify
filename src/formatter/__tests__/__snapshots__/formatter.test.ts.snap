// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic formatting 1`] = `
"{
    if (text) {
        test
    }
}"
`;

exports[`New line and tabs format 1`] = `
"{
    def variable = true
}"
`;
